test/unified_format.py

```diff
@@ -0,0 +1,997 @@
+        'modifiedCount': result.modified_count,
+        'upsertedCount': result.upserted_count,
+        'upsertedIds': upserted_ids}
+
+
+def parse_bulk_write_error_result(error):
+    write_result = BulkWriteResult(error.details, True)
+    return parse_bulk_write_result(write_result)
+
+
+class EventListenerUtil(CommandListener):
+    def __init__(self, observe_events, ignore_commands):
+        self._event_types = set(observe_events)
+        self._ignore_commands = _SENSITIVE_COMMANDS | set(ignore_commands)
+        self._ignore_commands.add('configurefailpoint')
+        self.results = []
+
+    def _observe_event(self, event):
+        if event.command_name.lower() not in self._ignore_commands:
+            self.results.append(event)
+
+    def started(self, event):
+        if 'commandStartedEvent' in self._event_types:
+            self._observe_event(event)
+
+    def succeeded(self, event):
+        if 'commandSucceededEvent' in self._event_types:
+            self._observe_event(event)
+
+    def failed(self, event):
+        if 'commandFailedEvent' in self._event_types:
+            self._observe_event(event)
+
+
+class EntityMapUtil(object):
+    """Utility class that implements an entity map as per the unified
+    test format specification."""
+    def __init__(self, test_class):
+        self._entities = {}
+        self._listeners = {}
+        self._session_lsids = {}
+        self._test_class = test_class
+
+    def __getitem__(self, item):
+        try:
+            return self._entities[item]
+        except KeyError:
+            self._test_class.fail('Could not find entity named %s in map' % (
+                item,))
+
+    def __setitem__(self, key, value):
+        if not isinstance(key, text_type):
+            self._test_class.fail(
+                'Expected entity name of type str, got %s' % (type(key)))
+
+        if key in self._entities:
+            self._test_class.fail('Entity named %s already in map' % (key,))
+
+        self._entities[key] = value
+
+    def _create_entity(self, entity_spec):
+        if len(entity_spec) != 1:
+            self._test_class.fail(
+                "Entity spec %s did not contain exactly one top-level key" % (
+                    entity_spec,))
+
+        entity_type, spec = next(iteritems(entity_spec))
+        if entity_type == 'client':
+            kwargs = {}
+            observe_events = spec.get('observeEvents', [])
+            ignore_commands = spec.get('ignoreCommandMonitoringEvents', [])
+            if len(observe_events) or len(ignore_commands):
+                ignore_commands = [cmd.lower() for cmd in ignore_commands]
+                listener = EventListenerUtil(observe_events, ignore_commands)
+                self._listeners[spec['id']] = listener
+                kwargs['event_listeners'] = [listener]
+            if client_context.is_mongos and spec.get('useMultipleMongoses'):
+                kwargs['host'] = client_context.mongos_seeds()
```

Comment:
Change this to `kwargs['h']` to fix the `host` errors. `rs_or_single_client` accepts a `h` argument, not `hosts`.

Reply:
Done.